src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -479,357 +487,304 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
 
                     var state = ReadState();
 
-                    return state switch
-                    {
-                        // if we already have a compilation, we must be already done!  This can happen if two
-                        // threads were waiting to build, and we came in after the other succeeded.
-                        FinalState finalState
-                            => finalState,
+                    // if we already have a compilation, we must be already done!  This can happen if two threads were
+                    // waiting to build, and we came in after the other succeeded.
+                    if (state is FinalState finalState)
+                        return finalState;
 
+                    // Transition from wherever we're currently at to the an in-progress state with some amount of 
+                    // translations to perform
+                    var expandedInProgressState = state switch
+                    {
                         // We've got nothing.  Build it from scratch :(
-                        null => await BuildFinalStateFromScratchAsync(
-                            compilationState,
-                            cancellationToken).ConfigureAwait(false),
-
-                        // We have a declaration compilation, use it to reconstruct the final compilation
-                        AllSyntaxTreesParsedState allSyntaxTreesParsedState => await FinalizeCompilationAsync(
-                            compilationState,
-                            allSyntaxTreesParsedState,
-                            compilationWithStaleGeneratedTrees: null,
-                            cancellationToken).ConfigureAwait(false),
+                        null => await BuildInProgressStateFromScratchAsync().ConfigureAwait(false),
 
                         // We must have an in progress compilation. Build off of that.
-                        InProgressState inProgressState => await BuildFinalStateFromInProgressStateAsync(
-                            compilationState,
-                            inProgressState,
-                            cancellationToken).ConfigureAwait(false),
+                        InProgressState inProgressState => inProgressState,
 
                         _ => throw ExceptionUtilities.UnexpectedValue(state.GetType()),
                     };
 
-                    async Task<FinalState> BuildFinalStateFromScratchAsync(
-                        SolutionCompilationState compilationState,
-                        CancellationToken cancellationToken)
+                    // Now do the final step of collapsing all pending transformations and then transition to the final state.
+                    var collapsedInProgressState = await CollapseInProgressStateAsync(expandedInProgressState).ConfigureAwait(false);
+                    return await FinalizeCompilationAsync(collapsedInProgressState).ConfigureAwait(false);
```

Comment:
finally, collapse the in-progress-state to its final form.  and then produce the final-state.